cpp/src/reshape/explode.cu

```diff
@@ -36,119 +36,207 @@
+                                              cudf::out_of_bounds_policy::NULLIFY,
+                                              stream,
+                                              mr)
+                                 ->release()[0])
+                   : std::make_unique<column>(sliced_child, stream, mr));
 
-  if (include_pos) {
+  if (position_array) {
+    size_type position_size = position_array->size();
     columns.insert(columns.begin() + explode_column_idx,
                    std::make_unique<column>(
-                     data_type(type_to_id<size_type>()), sliced_child.size(), pos.release()));
+                     data_type(type_to_id<size_type>()), position_size, position_array->release()));
   }
 
   return std::make_unique<table>(std::move(columns));
 }
 }  // namespace
 
-/**
- * @copydoc
- * cudf::explode(input_table,explode_column_idx,rmm::mr::device_memory_resource)
- *
- * @param stream CUDA stream used for device memory operations and kernel launches.
- */
 std::unique_ptr<table> explode(table_view const& input_table,
-                               size_type explode_column_idx,
-                               bool include_pos,
+                               size_type const explode_column_idx,
                                rmm::cuda_stream_view stream,
                                rmm::mr::device_memory_resource* mr)
 {
-  return type_dispatcher(input_table.column(explode_column_idx).type(),
-                         explode_functor{},
-                         input_table,
-                         explode_column_idx,
-                         include_pos,
-                         stream,
-                         mr);
+  lists_column_view explode_col{input_table.column(explode_column_idx)};
+  auto sliced_child = explode_col.get_sliced_child(stream);
+  rmm::device_uvector<size_type> gather_map(sliced_child.size(), stream);
+
+  // Sliced columns may require rebasing of the offsets.
+  auto offsets = explode_col.offsets_begin();
+  // offsets + 1 here to skip the 0th offset, which removes a - 1 operation later.
+  auto offsets_minus_one = thrust::make_transform_iterator(
+    thrust::next(offsets), [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
+  auto counting_iter = thrust::make_counting_iterator(0);
+
+  // This looks like an off-by-one bug, but what is going on here is that we need to reduce each
+  // result from `lower_bound` by 1 to build the correct gather map. This can be accomplished by
+  // skipping the first entry and using the result of `lower_bound` directly.
+  thrust::lower_bound(rmm::exec_policy(stream),
+                      offsets_minus_one,
+                      offsets_minus_one + explode_col.size(),
+                      counting_iter,
+                      counting_iter + gather_map.size(),
+                      gather_map.begin());
+
+  return build_table(input_table, explode_column_idx, sliced_child, gather_map, {}, {}, stream, mr);
```

Comment:
I take back what I said earlier. I think using `thrust::nullopt_t` here would make it clearer what those empty `{}` are.